pkg/statedb/reflector/k8s.go

```diff
@@ -0,0 +1,255 @@
+	src := stream.Buffer(
+		k8sEventObservable(s.ListerWatcher),
+		bufferSize,
+		waitTime,
+
+		// Buffer the events into a map, coalescing them by key.
+		func(buf buffer, ev cacheStoreEvent) buffer {
+			if buf == nil {
+				buf = make(buffer, bufferSize)
+			}
+			var entry entry
+			entry.deleted = ev.Type == cacheStoreDelete
+
+			var key string
+			if d, ok := ev.Obj.(cache.DeletedFinalStateUnknown); ok {
+				key = d.Key
+				entry.namespace, entry.name, _ = cache.SplitMetaNamespaceKey(d.Key)
+
+				entry.obj, ok = transform(d.Obj)
+				if !ok {
+					return buf
+				}
+			} else {
+				meta, err := meta.Accessor(ev.Obj)
+				if err != nil {
+					panic(fmt.Sprintf("%T internal error: meta.Accessor failed: %s", s, err))
+				}
+				entry.name = meta.GetName()
+				if ns := meta.GetNamespace(); ns != "" {
+					key = ns + "/" + meta.GetName()
+					entry.namespace = ns
+				} else {
+					key = meta.GetName()
+				}
+
+				var ok bool
+				entry.obj, ok = transform(ev.Obj)
+				if !ok {
+					return buf
+				}
+			}
+			buf[key] = entry
+			return buf
+		},
+	)
+
+	commitBuffer := func(buf buffer) {
+		txn := s.db.WriteTxn(s.Table)
+		defer txn.Commit()
+
+		for _, entry := range buf {
+			if !entry.deleted {
+				table.Insert(txn, entry.obj)
+			} else {
+				table.Delete(txn, entry.obj)
+			}
+		}
+	}
+
+	errs := make(chan error)
+	src.Observe(
+		ctx,
+		commitBuffer,
+		func(err error) {
+			errs <- err
+			close(errs)
+		},
+	)
+	return <-errs
+}
+
+// k8sEventObservable turns a ListerWatcher into an observable using the client-go's Reflector.
+func k8sEventObservable(lw cache.ListerWatcher) stream.Observable[cacheStoreEvent] {
+	return stream.FuncObservable[cacheStoreEvent](
+		func(ctx context.Context, next func(cacheStoreEvent), complete func(err error)) {
+			store := &cacheStoreListener{
+				onAdd: func(obj any) {
+					next(cacheStoreEvent{cacheStoreAdd, obj})
+				},
+				onUpdate: func(obj any) { next(cacheStoreEvent{cacheStoreUpdate, obj}) },
+				onDelete: func(obj any) { next(cacheStoreEvent{cacheStoreDelete, obj}) },
+			}
+			reflector := cache.NewReflector(lw, nil, store, 0)
+			go func() {
+				reflector.Run(ctx.Done())
+				complete(nil)
+			}()
+		})
+}
+
+const (
+	cacheStoreAdd = iota
+	cacheStoreUpdate
+	cacheStoreDelete
+)
+
+type cacheStoreEvent struct {
+	Type int
+	Obj  any
+}
+
+// cacheStoreListener implements the methods used by the cache reflector and
+// calls the given handlers for added, updated and deleted objects.
+type cacheStoreListener struct {
+	onAdd, onUpdate, onDelete func(any)
+}
+
+func (s *cacheStoreListener) Add(obj interface{}) error {
+	s.onAdd(obj)
+	return nil
+}
+
+func (s *cacheStoreListener) Update(obj interface{}) error {
+	s.onUpdate(obj)
+	return nil
+}
+
+func (s *cacheStoreListener) Delete(obj interface{}) error {
+	s.onDelete(obj)
+	return nil
+}
+
+func (s *cacheStoreListener) Replace(items []interface{}, resourceVersion string) error {
```

Comment:
Shouldn't `Replace` also delete all the items currently in the 'store', i.e. emit deletion callbacks? At least, that's the semantics in the comment on https://pkg.go.dev/k8s.io/client-go/tools/cache#Store

Reply:
Good question. I think we'd first need to answer the question of whether the reflector has full control over the table, or whether it's just one writer to it. My initial thinking was that we could have multiple writers (perhaps with a "Source" field in the object) and thus we can't just delete everything.

Reply:
Ok, refactored this so properly clear the table on `Replace`. It's required if `Watch` fails due to having fallen too much behind. But since I wanted to retain the ability to have one table populated from multiple reflectors/writers, I added a optional `QueryAll` parameter to limit deletions to subset of objects.

Reply:
Nice.